packages/components/src/modal/index.tsx

```diff
@@ -75,7 +87,13 @@ function UnforwardedModal(
 	const headingId = title
 		? `components-modal-header-${ instanceId }`
 		: aria.labelledby;
-	const focusOnMountRef = useFocusOnMount( focusOnMount );
+
+	// Modals should ignore the `Close` button which is the first focusable element.
+	// Remap `true` to select the next focusable element instead.
+	const focusOnMountRef = useFocusOnMount(
+		focusOnMount === true ? focusFirstNonCloseButtonElement : focusOnMount
```

Comment:
I wonder if we should throw an error or something because this code inadvertantly supports functions as props for the Modal component.

Reply:
what if the `focusOnMount` prop of the `Modal` component also supported passing a callback, same way as the hook?

Reply:
It did that until https://github.com/WordPress/gutenberg/pull/54296/commits/1f71011985d4822cb24c934802412c06a102f65e.

See https://github.com/WordPress/gutenberg/pull/54296#discussion_r1319691695

Reply:
I wanted to avoid opening too much APIs but I guess I may be on the minority here. It's fine.

Reply:
I see!

To recap: until this PR:
- passing `false` would disable the focus on mount functionality
- passing `true` would focus the `Modal`'s content wrapper
- passing `firstElement` would focus the first focusable element in terms of DOM order inside the `Modal` (ie. usually the close button)


What we could do, could be to:

- preserve same behaviour when passing `true` or `false`
- tweak the behaviour when passing `firstElement`, focusing the first element ignoring the header (as suggested in https://github.com/WordPress/gutenberg/pull/54296#discussion_r1319781088)
- optionally, also add support for passing a callback instead, thus enabling even more flexibility

We could also start without the callback option and do everything else, and add the callback only if we find it to be needed in real world usage.

Reply:
> add the callback only if we find it to be needed in real world usage.

I'd agree with that. 

> ...tweak the behaviour when passing firstElement, focusing the first element ignoring the header

Is it counter intuitive to have `firstElement` not actually focus the first element though?

Reply:
> tweak the behaviour when passing firstElement, focusing the first element ignoring the header
> We could specify that it will focus the first element inside the Modal's content (ie. excluding the header)?

Yes, that's my thinking too. Changing the "true" behavior is more disruptive and is harder to explain.

Reply:
I've updated the code. However the following comment still applies right?

> I wonder if we should throw an error or something because this code inadvertantly supports functions as props for the Modal component.

So if someone passes a `function` as `focusOnMount` to the `Modal` shall we coerce to `true` and then `console.warn` or just `throw`?

Reply:
Both the docs and the TypeScript types should suggest that the `focusOnMount` prop _doesn't_ support a function — in that sense, I don't think that any action is required. If folks want to try hacking around and pass a function, they're deciding to use the component in an unsupported way and they may incur in runtime errors.

The same would happen on any other component when passing prop values that are not supported